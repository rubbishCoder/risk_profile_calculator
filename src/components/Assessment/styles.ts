import { StyleSheet } from 'react-native';
import Colors from '../../common/Colors';
import RV from '../../common/utils/RV';

const styles = StyleSheet.create({
    textContainer: {
        backgroundColor: Colors.blue2,
        width: '100%',
        borderRadius: RV(10),
        padding: RV(16),
        marginBottom: RV(16)
    },
    footerContainer: {
        padding: RV(20),
        flexDirection: 'row',
        justifyContent: 'space-between'
    },
    prevButton: {
        flex: 1,
        padding: RV(16),
        marginRight: RV(16),
        justifyContent: 'center',
        alignItems: 'center',
        borderWidth: 1,
        borderColor: Colors.blue2,
        borderRadius: RV(8)
    },
    nextButton: {
        flex: 1,
        padding: RV(16),
        justifyContent: 'center',
        alignItems: 'center',
        borderWidth: 1,
        borderColor: Colors.blue2,
        borderRadius: RV(8),
        backgroundColor: Colors.blue2
    },
    prevButtonText: {
        fontSize: RV(16),
        textAlign: 'center',
        color: Colors.blue2,
        fontWeight: 'bold'
    },
    nextButtonText: {
        fontSize: RV(16),
        textAlign: 'center',
        color: Colors.white,
        fontWeight: 'bold'
    },
    questionCount: {
        paddingHorizontal: RV(16),
        paddingTop: RV(16),
        fontSize: RV(20),
        color: Colors.blue2,
        fontWeight: 'bold'
    },
    question: {
        fontSize: RV(20),
        color: Colors.white,
        fontWeight: 'bold',
    }
});

export default styles
