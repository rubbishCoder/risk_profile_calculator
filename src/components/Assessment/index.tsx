import { StackActions, useNavigation } from '@react-navigation/native';
import React from 'react';
import { SafeAreaView, ScrollView, Text, TouchableOpacity, View } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import Colors from '../../common/Colors';
import RadioButton from '../../common/components/RadioButton';
import RV from '../../common/utils/RV';
import { setQuestions, setRiskProfileScore, setSelectedQuestionIndex } from '../../reducers/questionsData';
import styles from './styles';

const Assessment = () => {
  const navigation = useNavigation<any>();
  const dispatch = useDispatch<any>();

  const { questions, selectedQuestionIndex } = useSelector((state: any) => ({
    questions: state.questionsData.questions,
    selectedQuestionIndex: state.questionsData.selectedQuestionIndex
  }))

  const selectedQuestion = questions[selectedQuestionIndex];

  const calculateRiskProfileScore = () => {
    const totalScore = questions?.reduce((acc: any, obj: any) => acc + obj?.options?.length, 0)
    const userScore = questions?.reduce((acc: any, obj: any) => acc + obj?.selectedOption?.score, 0)

    let riskProfileScore: any = (100 * userScore) / totalScore;
    riskProfileScore = parseInt(riskProfileScore);

    dispatch(setRiskProfileScore(riskProfileScore))

    navigation.dispatch(StackActions.replace('Results'));
  }

  return (
    <SafeAreaView style={{ flex: 1, backgroundColor: Colors.red, }}>
      {selectedQuestion &&
        <>
          <Text style={styles.questionCount}>{`Question ${selectedQuestionIndex + 1}/${questions?.length}`}</Text>
          <ScrollView contentContainerStyle={{ flexGrow: 1, padding: RV(16) }}>
            <Text style={styles.question}>{selectedQuestion?.question}</Text>
            {selectedQuestion?.options?.length > 0 &&
              selectedQuestion?.options?.map((d: any) => (
                <RadioButton
                  key={d?.value + d?.id}
                  text={d?.value}
                  onPress={() => {
                    let newQuestions = questions?.map((v: any, i: any) => {
                      let newData = v;
                      if (i === selectedQuestionIndex) {
                        newData = {
                          ...v,
                          selectedOption: d
                        }
                      }
                      return newData
                    })
                    dispatch(setQuestions(newQuestions))
                  }}
                  selected={selectedQuestion?.selectedOption?.id === d?.id} />
              ))}
          </ScrollView>
          <View style={styles.footerContainer}>
            {selectedQuestionIndex > 0 &&
              <TouchableOpacity style={styles.prevButton}
                onPress={() => {
                  dispatch(setSelectedQuestionIndex(selectedQuestionIndex - 1))
                }}>
                <Text style={styles.prevButtonText}>{`< Prev`}</Text>
              </TouchableOpacity>}
            <TouchableOpacity disabled={!selectedQuestion?.selectedOption}
              style={{
                ...styles.nextButton,
                marginLeft: selectedQuestionIndex > 0 ? 16 : undefined,
                opacity: !selectedQuestion?.selectedOption ? 0.5 : 1
              }}
              onPress={() => {
                if (selectedQuestionIndex === questions?.length - 1) {
                  calculateRiskProfileScore();
                } else {
                  dispatch(setSelectedQuestionIndex(selectedQuestionIndex + 1))
                }
              }}>
              <Text style={styles.nextButtonText}>{selectedQuestionIndex === questions?.length - 1 ? `Submit` : `Next >`}</Text>
            </TouchableOpacity>
          </View>
        </>}
    </SafeAreaView>
  )
}

export default Assessment;
