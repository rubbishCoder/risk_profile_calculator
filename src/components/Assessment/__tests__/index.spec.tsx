import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import * as Redux from 'react-redux';
import Assessment from '..';

let mockedNavigationDispatch = jest.fn();

jest.mock('@react-navigation/native', () => {
    return {
        useNavigation: () => ({
            dispatch: mockedNavigationDispatch
        }),
        StackActions: {
            replace: jest.fn().mockImplementation(x => ({ ...x, "type": "Navigation/REPLACE" }))
        }
    };
});

jest.mock('react-redux', () => {
    return {
        useDispatch: jest.fn(),
        useSelector: jest.fn()
    };
});

describe('Assessment', () => {
    beforeEach(() => {
        mockedNavigationDispatch.mockClear();
    });

    it('verify snapshot', () => {
        jest.spyOn(Redux, 'useSelector').mockReturnValue({
            questions: [
                {
                    questions: 'How are you?',
                    options: [
                        { id: 1, value: 'Bad', score: 1 },
                        { id: 2, value: 'Good', score: 2 },
                        { id: 3, value: 'Very Good', score: 3 },
                    ],
                    selectedOption: null
                }
            ],
            selectedQuestionIndex: 0
        });
        const AssessmentScreen = render(<Assessment />);
        const AssessmentScreenJson = AssessmentScreen.toJSON();
        expect(AssessmentScreenJson).toMatchSnapshot();
    });

    it('check next icon', () => {
        jest.spyOn(Redux, 'useSelector').mockReturnValue({
            questions: [
                {
                    questions: 'How are you?',
                    options: [
                        { id: 1, value: 'Bad', score: 1 },
                        { id: 2, value: 'Good', score: 2 },
                        { id: 3, value: 'Very Good', score: 3 },
                    ],
                    selectedOption: null
                },
                {
                    questions: 'What is you age?',
                    options: [
                        { id: 1, value: '20', score: 3 },
                        { id: 2, value: '40', score: 2 },
                        { id: 3, value: '60', score: 1 },
                    ],
                    selectedOption: null
                }
            ],
            selectedQuestionIndex: 0
        });
        const { getByText } = render(<Assessment />);
        expect(getByText('Next >')).toBeTruthy();
    });

    it('check prev and submit icon', () => {
        jest.spyOn(Redux, 'useSelector').mockReturnValue({
            questions: [
                {
                    questions: 'How are you?',
                    options: [
                        { id: 1, value: 'Bad', score: 1 },
                        { id: 2, value: 'Good', score: 2 },
                        { id: 3, value: 'Very Good', score: 3 },
                    ],
                    selectedOption: null
                },
                {
                    questions: 'What is you age?',
                    options: [
                        { id: 1, value: '20', score: 3 },
                        { id: 2, value: '40', score: 2 },
                        { id: 3, value: '60', score: 1 },
                    ],
                    selectedOption: null
                }
            ],
            selectedQuestionIndex: 1
        });
        const { getByText } = render(<Assessment />);
        expect(getByText('< Prev')).toBeTruthy();
        expect(getByText('Submit')).toBeTruthy();
    });

    it('check risk profile score calculation', () => {
        jest.spyOn(Redux, 'useSelector').mockReturnValue({
            questions: [
                {
                    questions: 'How are you?',
                    options: [
                        { id: 1, value: 'Bad', score: 1 },
                        { id: 2, value: 'Good', score: 2 },
                        { id: 3, value: 'Very Good', score: 3 },
                    ],
                    selectedOption: { id: 3, value: 'Very Good', score: 3 }
                },
                {
                    questions: 'What is you age?',
                    options: [
                        { id: 1, value: '20', score: 3 },
                        { id: 2, value: '40', score: 2 },
                        { id: 3, value: '60', score: 1 },
                    ],
                    selectedOption: { id: 3, value: '60', score: 1 }
                }
            ],
            selectedQuestionIndex: 1
        });
        let calculatedScore: any = 0;
        jest.spyOn(Redux, 'useDispatch').mockImplementation(() => jest.fn((x) => {
            calculatedScore = x.payload
            return x
        }));
        const { getByText } = render(<Assessment />);
        fireEvent.press(getByText('Submit'));
        expect(mockedNavigationDispatch).toHaveBeenCalled();
        expect(calculatedScore).toEqual(66);
    });
})