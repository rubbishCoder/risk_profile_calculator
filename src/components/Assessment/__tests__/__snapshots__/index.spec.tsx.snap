// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assessment verify snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#fe0133",
      "flex": 1,
    }
  }
>
  <Text
    style={
      {
        "color": "#ffffff",
        "fontSize": 29,
        "fontWeight": "bold",
        "paddingHorizontal": 23,
        "paddingTop": 23,
      }
    }
  >
    Question 1/1
  </Text>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
        "padding": 23,
      }
    }
  >
    <View>
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 29,
            "fontWeight": "bold",
            "marginBottom": 23,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderRadius": 12,
            "flexDirection": "row",
            "marginBottom": 23,
            "opacity": 1,
            "padding": 23,
            "shadowColor": "#3b393a",
            "shadowOffset": {
              "height": 4,
              "width": -2,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 3,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderRadius": 144,
              "borderWidth": 1,
              "height": 35,
              "justifyContent": "center",
              "width": 35,
            }
          }
          testID="SELECTED_ICON"
        />
        <Text
          style={
            {
              "color": "#000000",
              "fontSize": 29,
              "fontWeight": "bold",
              "paddingHorizontal": 23,
            }
          }
        >
          Bad
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderRadius": 12,
            "flexDirection": "row",
            "marginBottom": 23,
            "opacity": 1,
            "padding": 23,
            "shadowColor": "#3b393a",
            "shadowOffset": {
              "height": 4,
              "width": -2,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 3,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderRadius": 144,
              "borderWidth": 1,
              "height": 35,
              "justifyContent": "center",
              "width": 35,
            }
          }
          testID="SELECTED_ICON"
        />
        <Text
          style={
            {
              "color": "#000000",
              "fontSize": 29,
              "fontWeight": "bold",
              "paddingHorizontal": 23,
            }
          }
        >
          Good
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderRadius": 12,
            "flexDirection": "row",
            "marginBottom": 23,
            "opacity": 1,
            "padding": 23,
            "shadowColor": "#3b393a",
            "shadowOffset": {
              "height": 4,
              "width": -2,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 3,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderRadius": 144,
              "borderWidth": 1,
              "height": 35,
              "justifyContent": "center",
              "width": 35,
            }
          }
          testID="SELECTED_ICON"
        />
        <Text
          style={
            {
              "color": "#000000",
              "fontSize": 29,
              "fontWeight": "bold",
              "paddingHorizontal": 23,
            }
          }
        >
          Very Good
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 29,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#3b393a",
          "borderColor": "#3b393a",
          "borderRadius": 12,
          "borderWidth": 1,
          "flex": 1,
          "justifyContent": "center",
          "marginLeft": undefined,
          "opacity": 0.5,
          "padding": 23,
        }
      }
    >
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 23,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Submit
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
